src/k8s_upgrade.py

```diff
@@ -0,0 +1,497 @@
+                return
+        if not during_upgrade and self.charm.db_initialised and self.charm.is_db_service_ready():
+            self._upgrade.unit_state = UnitState.HEALTHY
+            self.charm.status.set_and_share_status(ActiveStatus())
+        if self.charm.unit.is_leader():
+            self._upgrade.reconcile_partition()
+
+        self._set_upgrade_status()
+
+    def _set_upgrade_status(self):
+        if self.charm.unit.is_leader():
+            self.charm.app.status = self._upgrade.app_status or ActiveStatus()
+        # Set/clear upgrade unit status if no other unit status - upgrade status for units should
+        # have the lowest priority.
+        if isinstance(self.charm.unit.status, ActiveStatus) or (
+            isinstance(self.charm.unit.status, BlockedStatus)
+            and self.charm.unit.status.message.startswith(
+                "Rollback with `juju refresh`. Pre-refresh check failed:"
+            )
+        ):
+            self.charm.status.set_and_share_status(
+                self._upgrade.get_unit_juju_status() or ActiveStatus()
+            )
+
+    def _on_upgrade_peer_relation_created(self, _) -> None:
+        """First time the relation is created, we save the revisions."""
+        if self.charm.unit.is_leader():
+            self._upgrade.set_versions_in_app_databag()
+
+    def _on_resume_upgrade_action(self, event: ActionEvent) -> None:
+        if not self.charm.unit.is_leader():
+            message = f"Must run action on leader unit. (e.g. `juju run {self.charm.app.name}/leader {RESUME_ACTION_NAME}`)"
+            logger.debug(f"Resume refresh failed: {message}")
+            event.fail(message)
+            return
+        if not self._upgrade or not self._upgrade.in_progress:
+            message = "No upgrade in progress"
+            logger.debug(f"Resume refresh failed: {message}")
+            event.fail(message)
+            return
+        self._upgrade.reconcile_partition(action_event=event)
+
+    def run_post_app_upgrade_task(self, event: EventBase):
+        """Runs the post upgrade check to verify that the cluster is healthy.
+
+        By deferring before setting unit state to HEALTHY, the user will either:
+            1. have to wait for the unit to resolve itself.
+            2. have to run the force-upgrade action (to upgrade the next unit).
+        """
+        logger.debug(
+            "Running post refresh checks to verify the deployment is not broken after refresh."
+        )
+        self.run_post_upgrade_checks(event, finished_whole_cluster=False)
+
+        if self._upgrade.unit_state != UnitState.HEALTHY:
+            logger.info(
+                f"Unit state is not healhy but {self._upgrade.unit_state}, not continuing post-refresh checks."
+            )
+            return
+
+        logger.debug("Cluster is healthy after refreshing unit %s", self.charm.unit.name)
+        self.charm.status.set_and_share_status(ActiveStatus())
```

Comment:
1. I don't think the status should be set to Active here (doesn't match VM)
2. I believe the log should also be removed (Even though it is present in the VM charm)

Reply:
1. I know and I think I added that initially because otherwise some outdated statuses would not get cleaned while the unit was healthy in case of a force upgrade. Again, I'll see what I can do for that:)
2. Ok I can do that